crypto/elliptic-curve.ts

```diff
@@ -703,3 +722,391 @@ function createCurveAffine({
+  let F = mod(E + D, p);
+  // H = Z^2
+  let H = mod(Z * Z, p);
+  // J =  F - 2 * H
+  let J = mod(F - 2n * H, p);
+
+  return {
+    x: mod((B - C - D) * J, p),
+    y: mod(F * (E - D), p),
+    z: mod(F * J, p),
+  };
+}
+
+function projectiveTwistedScale(
+  g: GroupProjective,
+  x: bigint | boolean[],
+  p: bigint,
+  a: bigint,
+  d: bigint
+) {
+  let bits = typeof x === 'bigint' ? bigIntToBits(x) : x;
+  let h = projectiveTwistedZero;
+  for (let bit of bits) {
+    if (bit) h = projectiveTwistedAdd(h, g, p, a, d);
+    g = projectiveTwistedDouble(g, p, a);
+  }
+  return h;
+}
+
+function projectiveFromAffineTwisted(g: GroupAffineTwisted): GroupProjective {
+  if (affineTwistedIsZero(g, p)) return projectiveTwistedZero;
+  return { x: g.x, y: g.y, z: 1n };
+}
+
+// The affine twisted curve with equation
+// a * x^2 + y^2 = 1 + d * x^2 * y^2
+// in projective coordinates is represented as
+// a * X^2 * Z^2 + Y^2 Z^2 = Z^4 + d * X^2 * Y^2
+// where x = X/Z, y = Y/Z, and Z ≠ 0
+function projectiveToAffineTwisted(
+  g: GroupProjective,
+  p: bigint
+): GroupAffineTwisted {
+  let z = g.z;
+  assert(
+    z !== 0n,
+    'projectiveToAffineTwisted: degenerate case, z must not be zero'
+  );
+  if (z === 1n) {
+    // any other normalized affine form
+    return { x: g.x, y: g.y };
+  } else {
+    let zinv = inverse(z, p)!; // we checked for z === 0, so inverse exists
+    // x/z
+    let x = mod(g.x * zinv, p);
+    // y/z
+    let y = mod(g.y * zinv, p);
+    return { x, y };
+  }
+}
+
+type AffineTwistedCurve = ReturnType<typeof createAffineTwistedCurve>;
+
+function affineTwistedIsZero(g: GroupAffineTwisted, p: bigint): boolean {
+  return mod(g.x, p) === 0n && mod(g.y, p) === 1n;
+}
+
+// auxiliary function to compute modular exponentiation (runs in O(log(exp)))
+function modPow(base: bigint, exp: bigint, mod: bigint) {
+  let result = 1n;
+  base = base % mod; // Reduce base initially
+
+  while (exp > 0n) {
+    if (exp % 2n === 1n) {
+      // If exponent is odd, multiply result
+      result = (result * base) % mod;
+    }
+    base = (base * base) % mod; // Square the base
+    exp = exp / 2n; // Halve the exponent
+  }
+
+  return result;
+}
+
+/** Creates twisted Edwards curves in affine cordinates of the form
+ * a * x^2 + y^2 = 1 + d * x^2 * y^2
+ * with a ≠ 0, d ≠ 0 and a ≠ d
+ *
+ * Warning: must be used only for curves without endomorphism, like edwards25519
+ */
+function createAffineTwistedCurve({
+  name,
+  modulus: p,
+  order,
+  cofactor,
+  generator,
+  a,
+  d,
+}: TwistedCurveParams) {
+  let hasCofactor = cofactor !== undefined && cofactor !== 1n;
+
+  const Field = createField(p);
+  const Scalar = createField(order);
+  const one = { ...generator };
+  const Endo = undefined; // for edwards25519
+
+  assert(a !== 0n, 'a must not be zero');
+  assert(d !== 0n, 'd must not be zero');
+  assert(a !== d, 'a must not be equal to d');
+  // Euler's criterion: d is square iff d^((p - 1) / 2) = 1 mod p
```

Comment:
`iff`

Reply:
math jargon 😄 https://en.wikipedia.org/wiki/If_and_only_if

Reply:
Ohhhhhhhhh

Reply:
I meant `iff` indeed, like if and only if 😄